### ISSUE TITLE
COPY with a file included in dockerignore issues `cannot stat`

### ISSUE THREAD

**Issue by eelstork** [docker/for-linux#44]
### Expected behavior

Using `COPY` (in a Dockerfile) with a file ignored via .dockerignore the expected behavior is that Docker would return a user friendly message indicating that the file cannot be copied because it matches an entry in .dockerignore.

### Actual behavior

Currently `docker build` will return a `cannot stat` error instead of providing a user friendly message. This is a problem in a team environment where the dev trying to use the COPY command may not be the same as the dev who created the .dockerignore.

### Information

  - Ubuntu Xenial
  - Docker version 17.05.0-ce, build 89658be

There is reaon to believe that same or similar symptoms may be observed on other platforms since this is a design bug.

### Steps to reproduce the behavior

1. cd to a new directory and create foo.txt
2. Create a .dockerignore file excluding foo.txt
3. Create a Dockerfile including COPY foot.txt .
4. Run `docker build`

**Comment by dnephin** [comment]
>  a user friendly message indicating that the file cannot be copied because it matches an entry in .dockerignore.

I don't think that is possible. The file is never included in the context, so the server would have no way of knowing if the file existed and was ignored, or if the file never existed on the client file system. Claiming that the file was not copied because it is in `.dockerignore` would be a lie.

**Comment by eelstork** [comment]
@dnephin very good point here. It looks to me that signalling to the user _possible_ reasons a COPY operation failed (for example, `File not found or not in context. Did you add this file to .dockerignore`) would suffice. 
Strictly, I think that the information contained in `.dockerignore` could be used to establish whether the target of a COPY operation would have been excluded from the context (whether it did exist or not) but I'm guessing approaching the problem this way isn't quite right.
I still think a more qualified error message would be helpful here.

**Comment by dnephin** [comment]
I think wrapping the error message with something more readable like "file not found or excluded by .dockerignore" would be a good idea.
